From 312f2dad813cf544ac36d01494f8a048060c7d3c Mon Sep 17 00:00:00 2001
From: sebres <info@sebres.de>
Date: Wed, 5 Nov 2025 19:13:22 +0100
Subject: [PATCH] amend (forgotten, unsaved LocalFileTimeToFileTime2 change)

---
 CPP/Windows/TimeUtils.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CPP/Windows/TimeUtils.h b/CPP/Windows/TimeUtils.h
index 4a9d0f2fa..bd5fc30af 100644
--- a/CPP/Windows/TimeUtils.h
+++ b/CPP/Windows/TimeUtils.h
@@ -137,7 +137,7 @@ inline bool PropVariant_SetFrom_DosTime(NWindows::NCOM::CPropVariant &prop, UInt
   FILETIME localFileTime, utc;
   if (!NWindows::NTime::DosTime_To_FileTime(dosTime, localFileTime))
     return false;
-  if (!LocalFileTimeToFileTime(&localFileTime, &utc))
+  if (!LocalFileTimeToFileTime2(&localFileTime, &utc))
     return false;
   prop.SetAsTimeFrom_FT_Prec(utc, k_PropVar_TimePrec_DOS);
   return true;
