From ea6457b4ba78a863837cd8afe63f091d55f5abfc Mon Sep 17 00:00:00 2001
From: Nobody <36154203+PopuriAO29@users.noreply.github.com>
Date: Sun, 14 Jan 2024 03:40:09 +0000
Subject: [PATCH]: add-extract-to-single-context-menu-option

---
diff --git a/CPP/7zip/UI/Explorer/ContextMenu.cpp b/CPP/7zip/UI/Explorer/ContextMenu.cpp
index fbecb700..d534492a 100644
--- a/CPP/7zip/UI/Explorer/ContextMenu.cpp
+++ b/CPP/7zip/UI/Explorer/ContextMenu.cpp
@@ -275,5 +275,6 @@ static const CContextMenuCommand g_Commands[] =
   CMD_REC( kExtractHere, "ExtractHere", IDS_CONTEXT_EXTRACT_HERE),
   CMD_REC( kExtractTo,   "ExtractTo",   IDS_CONTEXT_EXTRACT_TO),
+  CMD_REC( kExtractToSingle,        "ExtractToSingle",        IDS_CONTEXT_EXTRACT_TO_SINGLE),
   CMD_REC( kTest,        "Test",        IDS_CONTEXT_TEST),
   CMD_REC( kCompress,           "Compress",           IDS_CONTEXT_COMPRESS),
   CMD_REC( kCompressEmail,      "CompressEmail",      IDS_CONTEXT_COMPRESS_EMAIL),
@@ -923,6 +923,23 @@ Z7_COMWF_B CZipContextMenu::QueryContextMenu(HMENU hMenu, UINT indexMenu,
             MyInsertMenu(popupMenu, subIndex++, currentCommandID++, mainString, bitmap);
          }
        }
+ 
+        if ((contextMenuFlags & NContextMenuFlags::kExtractToSingle) != 0)
+        {
+          // Extract To Single
+          CCommandMapItem cmi;
+          AddCommand(kExtractToSingle, mainString, cmi);
+          /* if archive contain only one folder */
+          if (true)
+            {
+             cmi.Folder = baseFolder;
+            }
+          else
+            {
+             cmi.Folder = baseFolder + specFolder;
+            }
+        MyInsertMenu(popupMenu, subIndex++, currentCommandID++, mainString, bitmap);
+       }
       }
 
       if ((contextMenuFlags & NContextMenuFlags::kTest) != 0)
@@ -1310,5 +1327,6 @@ HRESULT CZipContextMenu::InvokeCommandCommon(const CCommandMapItem &cmi)
       case kExtractHere:
       case kExtractTo:
+      case kExtractToSingle:
       {
         if (_attribs.FirstDirIndex != -1)
         {
diff --git a/CPP/7zip/UI/Explorer/ContextMenu.h b/CPP/7zip/UI/Explorer/ContextMenu.h
index 668833f3..22863e90 100644
--- a/CPP/7zip/UI/Explorer/ContextMenu.h
+++ b/CPP/7zip/UI/Explorer/ContextMenu.h
@@ -79,5 +79,6 @@ public:
     kExtractHere,
     kExtractTo,
+    kExtractToSingle,
     kTest,
     kCompress,
     kCompressEmail,
diff --git a/CPP/7zip/UI/Explorer/ContextMenuFlags.h b/CPP/7zip/UI/Explorer/ContextMenuFlags.h
index a958d080..88051307 100644
--- a/CPP/7zip/UI/Explorer/ContextMenuFlags.h
+++ b/CPP/7zip/UI/Explorer/ContextMenuFlags.h
@@ -9,16 +9,17 @@ namespace NContextMenuFlags
   const UInt32 kExtractHere = 1 << 1;
   const UInt32 kExtractTo = 1 << 2;
+  const UInt32 kExtractToSingle = 1 << 3;
 
-  const UInt32 kTest = 1 << 4;
-  const UInt32 kOpen = 1 << 5;
-  const UInt32 kOpenAs = 1 << 6;
+  const UInt32 kTest = 1 << 5;
+  const UInt32 kOpen = 1 << 6;
+  const UInt32 kOpenAs = 1 << 7;
 
-  const UInt32 kCompress = 1 << 8;
-  const UInt32 kCompressTo7z = 1 << 9;
-  const UInt32 kCompressEmail = 1 << 10;
-  const UInt32 kCompressTo7zEmail = 1 << 11;
-  const UInt32 kCompressToZip = 1 << 12;
-  const UInt32 kCompressToZipEmail = 1 << 13;
+  const UInt32 kCompress = 1 << 9;
+  const UInt32 kCompressTo7z = 1 << 10;
+  const UInt32 kCompressEmail = 1 << 11;
+  const UInt32 kCompressTo7zEmail = 1 << 12;
+  const UInt32 kCompressToZip = 1 << 13;
+  const UInt32 kCompressToZipEmail = 1 << 14;
 
   const UInt32 kCRC_Cascaded = (UInt32)1 << 30;
   const UInt32 kCRC = (UInt32)1 << 31;
diff --git a/CPP/7zip/UI/Explorer/resource.h b/CPP/7zip/UI/Explorer/resource.h
index 88ff30b0..88c7bbef 100644
--- a/CPP/7zip/UI/Explorer/resource.h
+++ b/CPP/7zip/UI/Explorer/resource.h
@@ -6,9 +6,11 @@
 #define IDS_CONTEXT_TEST                2325
 #define IDS_CONTEXT_EXTRACT_HERE        2326
 #define IDS_CONTEXT_EXTRACT_TO          2327
-#define IDS_CONTEXT_COMPRESS_TO         2329
-#define IDS_CONTEXT_COMPRESS_EMAIL      2330
-#define IDS_CONTEXT_COMPRESS_TO_EMAIL   2331
+#define IDS_CONTEXT_EXTRACT_TO_SINGLE   2329
+#define IDS_CONTEXT_COMPRESS_TO         2330
+#define IDS_CONTEXT_COMPRESS_EMAIL      2331
+#define IDS_CONTEXT_COMPRESS_TO_EMAIL   2332
+
 
 #define IDS_SELECT_FILES                3015
 
diff --git a/CPP/7zip/UI/Explorer/resource2.rc b/CPP/7zip/UI/Explorer/resource2.rc
index c5cc0524..eb06381f 100644
--- a/CPP/7zip/UI/Explorer/resource2.rc
+++ b/CPP/7zip/UI/Explorer/resource2.rc
@@ -10,6 +10,7 @@ BEGIN
   IDS_CONTEXT_TEST        "Test archive"
   IDS_CONTEXT_EXTRACT_HERE "Extract Here"
   IDS_CONTEXT_EXTRACT_TO  "Extract to {0}"
+  IDS_CONTEXT_EXTRACT_TO_SINGLE "Extract to single folder"
   IDS_CONTEXT_COMPRESS_TO "Add to {0}"
   IDS_CONTEXT_COMPRESS_EMAIL "Compress and email..."
   IDS_CONTEXT_COMPRESS_TO_EMAIL "Compress to {0} and email"
diff --git a/CPP/7zip/UI/FileManager/MenuPage.cpp b/CPP/7zip/UI/FileManager/MenuPage.cpp
index 09e18d2f..c0452e50 100644
--- a/CPP/7zip/UI/FileManager/MenuPage.cpp
+++ b/CPP/7zip/UI/FileManager/MenuPage.cpp
@@ -56,5 +56,6 @@ static const CContextMenuItem kMenuItems[] =
   { IDS_CONTEXT_EXTRACT_HERE, kExtractHere },
   { IDS_CONTEXT_EXTRACT_TO, kExtractTo },
+  { IDS_CONTEXT_EXTRACT_TO_SINGLE, kExtractToSingle },
 
   { IDS_CONTEXT_TEST, kTest },
 

