From 6ae53966535af934b3ca276d82e15b29a2468d85 Mon Sep 17 00:00:00 2001
From: ozone10 <positronix10@gmail.com>
Date: Fri, 4 Apr 2025 17:17:23 +0200
Subject: [PATCH] Tweaks

---
 DarkMode/DarkModeSubclass.cpp | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/DarkMode/DarkModeSubclass.cpp b/DarkMode/DarkModeSubclass.cpp
index 2417ca5f3..6dab37a56 100644
--- a/DarkMode/DarkModeSubclass.cpp
+++ b/DarkMode/DarkModeSubclass.cpp
@@ -81,13 +81,13 @@ static std::wstring getWndClassName(HWND hWnd)
 {
 	constexpr int strLen = 32;
 	std::wstring className(strLen, 0);
-	className.resize(::GetClassName(hWnd, &className[0], strLen));
+	className.resize(::GetClassNameW(hWnd, className.data(), strLen));
 	return className;
 }
 
-static bool cmpWndClassName(HWND hWnd, std::wstring classNameToCmp)
+static bool cmpWndClassName(HWND hWnd, const wchar_t* classNameToCmp)
 {
-	return getWndClassName(hWnd) == classNameToCmp;
+	return (getWndClassName(hWnd) == classNameToCmp);
 }
 
 static std::wstring getIniPath(std::wstring iniFilename)
@@ -931,11 +931,6 @@ namespace DarkMode
 			closeTheme();
 		}
 
-		const wchar_t *getThemeClass() const
-		{
-			return _themeClass;
-		}
-
 		bool ensureTheme(HWND hWnd)
 		{
 			if (!_hTheme && _themeClass)
@@ -1914,7 +1909,7 @@ namespace DarkMode
 		bool hasFocus = false;
 
 		::SelectObject(hdc, reinterpret_cast<HFONT>(::SendMessage(hWnd, WM_GETFONT, 0, 0)));
-		//::SetBkMode(hdc, TRANSPARENT);
+		::SetBkMode(hdc, TRANSPARENT); // for non-theme DrawText
 
 		RECT rcArrow{ cbi.rcButton };
 		rcArrow.left -= 1;
@@ -2984,7 +2979,7 @@ namespace DarkMode
 			DWORD cchText = 0;
 			cchText = LOWORD(::SendMessage(hWnd, SB_GETTEXTLENGTH, i, 0));
 			str.resize(cchText + 1); // technically the std::wstring might not have an internal null character at the end of the buffer, so add one
-			LRESULT lr = ::SendMessage(hWnd, SB_GETTEXT, i, reinterpret_cast<LPARAM>(&str[0]));
+			LRESULT lr = ::SendMessage(hWnd, SB_GETTEXT, i, reinterpret_cast<LPARAM>(str.data()));
 			str.resize(cchText); // remove the extra NULL character
 			bool ownerDraw = false;
 			if (cchText == 0 && (lr & ~(SBT_NOBORDERS | SBT_POPOUT | SBT_RTLREADING)) != 0)
