From 137609635cc0bb225ffff0decd9b329ad8e8cac9 Mon Sep 17 00:00:00 2001
From: ozone10 <positronix10@gmail.com>
Date: Mon, 28 Apr 2025 22:01:44 +0200
Subject: [PATCH] Refactor

---
 DarkMode/DarkModeSubclass.cpp | 624 ++++++++++++++++++----------------
 DarkMode/DarkModeSubclass.h   | 112 +++---
 2 files changed, 402 insertions(+), 334 deletions(-)

diff --git a/DarkMode/DarkModeSubclass.cpp b/DarkMode/DarkModeSubclass.cpp
index f9a1b29e9..208174170 100644
--- a/DarkMode/DarkModeSubclass.cpp
+++ b/DarkMode/DarkModeSubclass.cpp
@@ -176,6 +176,13 @@ namespace DarkMode
 		settingChange   = 16
 	};
 
+	struct DarkModeParams
+	{
+		const wchar_t* _themeClassName = nullptr;
+		bool _subclass = false;
+		bool _theme = false;
+	};
+
 	struct Brushes
 	{
 		HBRUSH background = nullptr;
@@ -412,50 +419,50 @@ namespace DarkMode
 		::GetSysColor(COLOR_GRAYTEXT)       // disabledEdgeColor
 	};
 
-	static ColorTone g_colorToneChoice = DarkMode::ColorTone::blackTone;
+	static ColorTone g_colorToneChoice = DarkMode::ColorTone::black;
 
 	void setDarkCustomColors(ColorTone colorTone)
 	{
 		switch (colorTone)
 		{
-			case DarkMode::ColorTone::redTone:
+			case DarkMode::ColorTone::red:
 			{
 				darkCustomizedColors = darkRedColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::greenTone:
+			case DarkMode::ColorTone::green:
 			{
 				darkCustomizedColors = darkGreenColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::blueTone:
+			case DarkMode::ColorTone::blue:
 			{
 				darkCustomizedColors = darkBlueColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::purpleTone:
+			case DarkMode::ColorTone::purple:
 			{
 				darkCustomizedColors = darkPurpleColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::cyanTone:
+			case DarkMode::ColorTone::cyan:
 			{
 				darkCustomizedColors = darkCyanColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::oliveTone:
+			case DarkMode::ColorTone::olive:
 			{
 				darkCustomizedColors = darkOliveColors;
 				break;
 			}
 
-			case DarkMode::ColorTone::customizedTone:
-			case DarkMode::ColorTone::blackTone:
+			case DarkMode::ColorTone::customized:
+			case DarkMode::ColorTone::black:
 			{
 				darkCustomizedColors = darkColors;
 				break;
@@ -485,6 +492,11 @@ namespace DarkMode
 
 	static Theme tDefault(darkCustomizedColors);
 
+	static Theme& getTheme()
+	{
+		return tDefault;
+	}
+
 	static ColorsView darkColorsView{
 		RGB(41, 49, 52),      // background
 		RGB(224, 226, 228),   // text
@@ -505,11 +517,6 @@ namespace DarkMode
 		RGB(229, 229, 229)    // header divider
 	};
 
-	static Theme& getTheme()
-	{
-		return tDefault;
-	}
-
 	struct BrushesAndPensView
 	{
 		HBRUSH background = nullptr;
@@ -517,7 +524,7 @@ namespace DarkMode
 		HBRUSH headerBackground = nullptr;
 		HBRUSH headerHotBackground = nullptr;
 
-		HPEN edge = nullptr;
+		HPEN headerEdge = nullptr;
 
 		BrushesAndPensView(const ColorsView& colors)
 			: background(::CreateSolidBrush(colors.background))
@@ -525,7 +532,7 @@ namespace DarkMode
 			, headerBackground(::CreateSolidBrush(colors.headerBackground))
 			, headerHotBackground(::CreateSolidBrush(colors.headerHotBackground))
 
-			, edge(::CreatePen(PS_SOLID, 1, colors.headerEdge))
+			, headerEdge(::CreatePen(PS_SOLID, 1, colors.headerEdge))
 		{}
 
 		~BrushesAndPensView()
@@ -535,7 +542,7 @@ namespace DarkMode
 			::DeleteObject(headerBackground);       headerBackground = nullptr;
 			::DeleteObject(headerHotBackground);    headerHotBackground = nullptr;
 
-			::DeleteObject(edge);                   edge = nullptr;
+			::DeleteObject(headerEdge);             headerEdge = nullptr;
 		}
 
 		void change(const ColorsView& colors)
@@ -550,9 +557,9 @@ namespace DarkMode
 			headerBackground = ::CreateSolidBrush(colors.headerBackground);
 			headerHotBackground = ::CreateSolidBrush(colors.headerHotBackground);
 
-			::DeleteObject(edge);
+			::DeleteObject(headerEdge);
 
-			edge = ::CreatePen(PS_SOLID, 1, colors.headerEdge);
+			headerEdge = ::CreatePen(PS_SOLID, 1, colors.headerEdge);
 		}
 	};
 
@@ -590,12 +597,105 @@ namespace DarkMode
 		return tView;
 	}
 
+	inline static COLORREF setNewColor(COLORREF* clrOld, COLORREF clrNew)
+	{
+		auto clrTmp = *clrOld;
+		*clrOld = clrNew;
+		return clrTmp;
+	}
+
+	COLORREF setBackgroundColor(COLORREF clrNew)        { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.background, clrNew); }
+	COLORREF setCtrlBackgroundColor(COLORREF clrNew)    { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.ctrlBackground, clrNew); }
+	COLORREF setHotBackgroundColor(COLORREF clrNew)     { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.hotBackground, clrNew); }
+	COLORREF setDlgBackgroundColor(COLORREF clrNew)     { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.dlgBackground, clrNew); }
+	COLORREF setErrorBackgroundColor(COLORREF clrNew)   { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.errorBackground, clrNew); }
+	COLORREF setTextColor(COLORREF clrNew)              { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.text, clrNew); }
+	COLORREF setDarkerTextColor(COLORREF clrNew)        { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.darkerText, clrNew); }
+	COLORREF setDisabledTextColor(COLORREF clrNew)      { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.disabledText, clrNew); }
+	COLORREF setLinkTextColor(COLORREF clrNew)          { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.linkText, clrNew); }
+	COLORREF setEdgeColor(COLORREF clrNew)              { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.edge, clrNew); }
+	COLORREF setHotEdgeColor(COLORREF clrNew)           { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.hotEdge, clrNew); }
+	COLORREF setDisabledEdgeColor(COLORREF clrNew)      { return DarkMode::setNewColor(&DarkMode::getTheme()._colors.disabledEdge, clrNew); }
+
+	void changeCustomTheme(const Colors& colors)
+	{
+		tDefault.change(colors);
+	}
+
+	void updateBrushesAndPens()
+	{
+		DarkMode::getTheme()._brushes.change(DarkMode::getTheme()._colors);
+		DarkMode::getTheme()._pens.change(DarkMode::getTheme()._colors);
+	}
+
+	COLORREF getBackgroundColor()         { return getTheme()._colors.background; }
+	COLORREF getCtrlBackgroundColor()     { return getTheme()._colors.ctrlBackground; }
+	COLORREF getHotBackgroundColor()      { return getTheme()._colors.hotBackground; }
+	COLORREF getDlgBackgroundColor()      { return getTheme()._colors.dlgBackground; }
+	COLORREF getErrorBackgroundColor()    { return getTheme()._colors.errorBackground; }
+	COLORREF getTextColor()               { return getTheme()._colors.text; }
+	COLORREF getDarkerTextColor()         { return getTheme()._colors.darkerText; }
+	COLORREF getDisabledTextColor()       { return getTheme()._colors.disabledText; }
+	COLORREF getLinkTextColor()           { return getTheme()._colors.linkText; }
+	COLORREF getEdgeColor()               { return getTheme()._colors.edge; }
+	COLORREF getHotEdgeColor()            { return getTheme()._colors.hotEdge; }
+	COLORREF getDisabledEdgeColor()       { return getTheme()._colors.disabledEdge; }
+
+	HBRUSH getBackgroundBrush()           { return getTheme()._brushes.background; }
+	HBRUSH getCtrlBackgroundBrush()       { return getTheme()._brushes.ctrlBackground; }
+	HBRUSH getHotBackgroundBrush()        { return getTheme()._brushes.hotBackground; }
+	HBRUSH getDlgBackgroundBrush()        { return getTheme()._brushes.dlgBackground; }
+	HBRUSH getErrorBackgroundBrush()      { return getTheme()._brushes.errorBackground; }
+
+	HBRUSH getEdgeBrush()                 { return getTheme()._brushes.edge; }
+	HBRUSH getHotEdgeBrush()              { return getTheme()._brushes.hotEdge; }
+	HBRUSH getDisabledEdgeBrush()         { return getTheme()._brushes.disabledEdge; }
+
+	HPEN getDarkerTextPen()               { return getTheme()._pens.darkerTextPen; }
+	HPEN getEdgePen()                     { return getTheme()._pens.edgePen; }
+	HPEN getHotEdgePen()                  { return getTheme()._pens.hotEdgePen; }
+	HPEN getDisabledEdgePen()             { return getTheme()._pens.disabledEdgePen; }
+
+	COLORREF setViewBackgroundColor(COLORREF clrNew)        { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.background, clrNew); }
+	COLORREF setViewTextColor(COLORREF clrNew)              { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.text, clrNew); }
+	COLORREF setViewGridlinesColor(COLORREF clrNew)         { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.gridlines, clrNew); }
+
+	COLORREF setHeaderBackgroundColor(COLORREF clrNew)      { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.headerBackground, clrNew); }
+	COLORREF setHeaderHotBackgroundColor(COLORREF clrNew)   { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.headerHotBackground, clrNew); }
+	COLORREF setHeaderTextColor(COLORREF clrNew)            { return DarkMode::setNewColor(&DarkMode::getThemeView()._clrView.headerText, clrNew); }
+
+	void updateBrushesAndPensView()
+	{
+		DarkMode::getThemeView().updateBrushesAndPens();
+	}
+
+	COLORREF getViewBackgroundColor()       { return DarkMode::getThemeView()._clrView.background; }
+	COLORREF getViewTextColor()             { return DarkMode::getThemeView()._clrView.text; }
+	COLORREF getViewGridlinesColor()        { return DarkMode::getThemeView()._clrView.gridlines; }
+
+	COLORREF getHeaderBackgroundColor()     { return DarkMode::getThemeView()._clrView.headerBackground; }
+	COLORREF getHeaderHotBackgroundColor()  { return DarkMode::getThemeView()._clrView.headerHotBackground; }
+	COLORREF getHeaderTextColor()           { return DarkMode::getThemeView()._clrView.headerText; }
+
+	HBRUSH getViewBackgroundBrush()         { return DarkMode::getThemeView()._hbrPnView.background; }
+	HBRUSH getViewGridlinesBrush()          { return DarkMode::getThemeView()._hbrPnView.gridlines; }
+
+	HBRUSH getHeaderBackgroundBrush()       { return DarkMode::getThemeView()._hbrPnView.headerBackground; }
+	HBRUSH getHeaderHotBackgroundBrush()    { return DarkMode::getThemeView()._hbrPnView.headerHotBackground; }
+
+	HPEN getHeaderEdgePen()                 { return DarkMode::getThemeView()._hbrPnView.headerEdge; }
+
+	static TreeViewStyle g_treeViewStyle = TreeViewStyle::classic;
+
 	static bool g_useDarkMode = true;
 	static bool g_enableWindowsMode = false;
 
 	static auto g_mica = DWMSBT_AUTO;
 	static bool g_micaExtend = false;
 
+	static bool g_isInit = false;
+	static bool g_isInitExperimental = false;
+
 	static void initOptions()
 	{
 		std::wstring iniPath = getIniPath(g_iniName);
@@ -697,7 +797,7 @@ namespace DarkMode
 			setClrFromIni(sectionColors, L"backgroundInteractive", iniPath, &DarkMode::getTheme()._colors.ctrlBackground);
 			setClrFromIni(sectionColors, L"backgroundHot", iniPath, &DarkMode::getTheme()._colors.hotBackground);
 			setClrFromIni(sectionColors, L"backgroundDlg", iniPath, &DarkMode::getTheme()._colors.dlgBackground);
-			//setClrFromIni(sectionColors, L"backgroundError", iniPath, &DarkMode::getTheme()._colors.errorBackground);
+			setClrFromIni(sectionColors, L"backgroundError", iniPath, &DarkMode::getTheme()._colors.errorBackground);
 
 			setClrFromIni(sectionColors, L"text", iniPath, &DarkMode::getTheme()._colors.text);
 			setClrFromIni(sectionColors, L"textItem", iniPath, &DarkMode::getTheme()._colors.darkerText);
@@ -708,15 +808,30 @@ namespace DarkMode
 			setClrFromIni(sectionColors, L"edgeHot", iniPath, &DarkMode::getTheme()._colors.hotEdge);
 			setClrFromIni(sectionColors, L"edgeDisabled", iniPath, &DarkMode::getTheme()._colors.disabledEdge);
 
-			DarkMode::getTheme()._brushes.change(DarkMode::getTheme()._colors);
-			DarkMode::getTheme()._pens.change(DarkMode::getTheme()._colors);
-
-			DarkMode::getThemeView().updateBrushesAndPens();
+			DarkMode::updateBrushesAndPens();
+			DarkMode::updateBrushesAndPensView();
 		}
 	}
 
-	static bool g_isInit = false;
-	static bool g_isInitExperimental = false;
+	static void initExperimentalDarkMode()
+	{
+		::InitDarkMode();
+	}
+
+	static void setDarkMode(bool useDark, bool fixDarkScrollbar)
+	{
+		::SetDarkMode(useDark, fixDarkScrollbar);
+	}
+
+	static bool allowDarkModeForWindow(HWND hWnd, bool allow)
+	{
+		return ::AllowDarkModeForWindow(hWnd, allow);
+	}
+
+	static void setTitleBarThemeColor(HWND hWnd)
+	{
+		::RefreshTitleBarThemeColor(hWnd);
+	}
 
 	void initDarkMode()
 	{
@@ -728,7 +843,7 @@ namespace DarkMode
 				g_isInitExperimental = true;
 			}
 
-			initOptions();
+			DarkMode::initOptions();
 
 			DarkMode::calculateTreeViewStyle();
 			DarkMode::setDarkMode(g_useDarkMode, true);
@@ -776,77 +891,9 @@ namespace DarkMode
 		return GetWindowsBuildNumber();
 	}
 
-	static TreeViewStyle g_treeViewStylePrev = TreeViewStyle::classic;
-	static TreeViewStyle g_treeViewStyle = TreeViewStyle::classic;
-	static COLORREF g_treeViewBg = RGB(41, 49, 52);
-	static double g_lightnessTreeView = 50.0;
-
-	// adapted from https://stackoverflow.com/a/56678483
-	double calculatePerceivedLightness(COLORREF c)
-	{
-		auto linearValue = [](double colorChannel) -> double {
-			colorChannel /= 255.0;
-			if (colorChannel <= 0.04045)
-				return colorChannel / 12.92;
-			return std::pow(((colorChannel + 0.055) / 1.055), 2.4);
-			};
-
-		double r = linearValue(static_cast<double>(GetRValue(c)));
-		double g = linearValue(static_cast<double>(GetGValue(c)));
-		double b = linearValue(static_cast<double>(GetBValue(c)));
-
-		double luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;
-
-		double lightness = (luminance <= 216.0 / 24389.0) ? (luminance * 24389.0 / 27.0) : (std::pow(luminance, (1.0 / 3.0)) * 116.0 - 16.0);
-		return lightness;
-	}
-
-	COLORREF getBackgroundColor()         { return getTheme()._colors.background; }
-	COLORREF getCtrlBackgroundColor()     { return getTheme()._colors.ctrlBackground; }
-	COLORREF getHotBackgroundColor()      { return getTheme()._colors.hotBackground; }
-	COLORREF getDlgBackgroundColor()      { return getTheme()._colors.dlgBackground; }
-	COLORREF getErrorBackgroundColor()    { return getTheme()._colors.errorBackground; }
-	COLORREF getTextColor()               { return getTheme()._colors.text; }
-	COLORREF getDarkerTextColor()         { return getTheme()._colors.darkerText; }
-	COLORREF getDisabledTextColor()       { return getTheme()._colors.disabledText; }
-	COLORREF getLinkTextColor()           { return getTheme()._colors.linkText; }
-	COLORREF getEdgeColor()               { return getTheme()._colors.edge; }
-	COLORREF getHotEdgeColor()            { return getTheme()._colors.hotEdge; }
-	COLORREF getDisabledEdgeColor()       { return getTheme()._colors.disabledEdge; }
-
-	HBRUSH getBackgroundBrush()           { return getTheme()._brushes.background; }
-	HBRUSH getCtrlBackgroundBrush()       { return getTheme()._brushes.ctrlBackground; }
-	HBRUSH getHotBackgroundBrush()        { return getTheme()._brushes.hotBackground; }
-	HBRUSH getDlgBackgroundBrush()        { return getTheme()._brushes.dlgBackground; }
-	HBRUSH getErrorBackgroundBrush()      { return getTheme()._brushes.errorBackground; }
-
-	HBRUSH getEdgeBrush()                 { return getTheme()._brushes.edge; }
-	HBRUSH getHotEdgeBrush()              { return getTheme()._brushes.hotEdge; }
-	HBRUSH getDisabledEdgeBrush()         { return getTheme()._brushes.disabledEdge; }
-
-	HPEN getDarkerTextPen()               { return getTheme()._pens.darkerTextPen; }
-	HPEN getEdgePen()                     { return getTheme()._pens.edgePen; }
-	HPEN getHotEdgePen()                  { return getTheme()._pens.hotEdgePen; }
-	HPEN getDisabledEdgePen()             { return getTheme()._pens.disabledEdgePen; }
-
-	void changeCustomTheme(const Colors& colors)
-	{
-		tDefault.change(colors);
-	}
-
-	COLORREF getViewBackgroundColor()       { return DarkMode::getThemeView()._clrView.background; }
-	COLORREF getViewTextColor()             { return DarkMode::getThemeView()._clrView.text; }
-	COLORREF getViewGridlinesColor()        { return DarkMode::getThemeView()._clrView.gridlines; }
-	COLORREF getHeaderTextColor()           { return DarkMode::getThemeView()._clrView.headerText; }
-	HBRUSH getViewBackgroundBrush()         { return DarkMode::getThemeView()._hbrPnView.background; }
-	HBRUSH getViewGridlinesBrush()          { return DarkMode::getThemeView()._hbrPnView.gridlines; }
-	HBRUSH getHeaderBackgroundBrush()       { return DarkMode::getThemeView()._hbrPnView.headerBackground; }
-	HBRUSH getHeaderHotBackgroundBrush()    { return DarkMode::getThemeView()._hbrPnView.headerHotBackground; }
-	HPEN getHeaderEdgePen()                 { return DarkMode::getThemeView()._hbrPnView.edge; }
-
 	bool handleSettingChange(LPARAM lParam)
 	{
-		if (DarkMode::isExperimentalSupported() && IsColorSchemeChangeMessage(lParam))
+		if (DarkMode::isExperimentalSupported() && ::IsColorSchemeChangeMessage(lParam))
 		{
 			// ShouldAppsUseDarkMode() is not reliable from 1903+, use NppDarkMode::isDarkModeReg() instead
 			bool isDarkModeUsed = DarkMode::isDarkModeReg() && !IsHighContrast();
@@ -883,31 +930,6 @@ namespace DarkMode
 
 	// from DarkMode.h
 
-	void initExperimentalDarkMode()
-	{
-		::InitDarkMode();
-	}
-
-	void setDarkMode(bool useDark, bool fixDarkScrollbar)
-	{
-		::SetDarkMode(useDark, fixDarkScrollbar);
-	}
-
-	void allowDarkModeForApp(bool allow)
-	{
-		::AllowDarkModeForApp(allow);
-	}
-
-	bool allowDarkModeForWindow(HWND hWnd, bool allow)
-	{
-		return ::AllowDarkModeForWindow(hWnd, allow);
-	}
-
-	void setTitleBarThemeColor(HWND hWnd)
-	{
-		::RefreshTitleBarThemeColor(hWnd);
-	}
-
 	void setSysColor(int nIndex, COLORREF color)
 	{
 		::SetMySysColor(nIndex, color);
@@ -2122,7 +2144,7 @@ namespace DarkMode
 		DarkMode::subclassTabControlUpDown(hWnd);
 	}
 
-	void subclassTabControl(HWND hWnd, DarkModeParams p)
+	static void subclassTabControl(HWND hWnd, DarkModeParams p)
 	{
 		if (p._subclass)
 		{
@@ -2653,7 +2675,7 @@ namespace DarkMode
 		return ::DefSubclassProc(hWnd, uMsg, wParam, lParam);
 	}
 
-	static void subclassComboboxEx(HWND hWnd)
+	void subclassComboboxExControl(HWND hWnd)
 	{
 		if (::GetWindowSubclass(hWnd, ComboboxExSubclass, g_comboboxExSubclassID, nullptr) == FALSE)
 		{
@@ -2661,11 +2683,11 @@ namespace DarkMode
 		}
 	}
 
-	void subclassComboboxEx(HWND hWnd, DarkModeParams p)
+	static void subclassComboboxExControl(HWND hWnd, DarkModeParams p)
 	{
 		if (p._subclass)
 		{
-			subclassComboboxEx(hWnd);
+			DarkMode::subclassComboboxExControl(hWnd);
 		}
 	}
 
@@ -2684,8 +2706,8 @@ namespace DarkMode
 		{
 			case WM_NCDESTROY:
 			{
-				DarkMode::unhookSysColor();
 				::RemoveWindowSubclass(hWnd, ListViewSubclass, uIdSubclass);
+				DarkMode::unhookSysColor();
 				break;
 			}
 
@@ -3335,7 +3357,7 @@ namespace DarkMode
 		}
 	}
 
-	void subclassStatusBarControl(HWND hWnd, DarkModeParams p)
+	static void subclassStatusBarControl(HWND hWnd, DarkModeParams p)
 	{
 		if (p._subclass)
 		{
@@ -3507,6 +3529,7 @@ namespace DarkMode
 						pProgressBarData->_iStateID = PBFS_ERROR; // red
 						break;
 					}
+
 					case PBST_PAUSED:
 					{
 						pProgressBarData->_iStateID = PBFS_PAUSED; // yellow
@@ -3528,7 +3551,7 @@ namespace DarkMode
 		}
 	}
 
-	void subclassProgressBarControl(HWND hWnd, DarkModeParams p)
+	static void subclassProgressBarControl(HWND hWnd, DarkModeParams p)
 	{
 		if (p._subclass)
 		{
@@ -3602,7 +3625,7 @@ namespace DarkMode
 		}
 	}
 
-	void subclassStaticText(HWND hWnd, DarkModeParams p)
+	static void subclassStaticText(HWND hWnd, DarkModeParams p)
 	{
 		if (p._subclass)
 		{
@@ -3610,144 +3633,7 @@ namespace DarkMode
 		}
 	}
 
-	void autoSubclassAndThemeChildControls(HWND hwndParent, bool subclass, bool theme)
-	{
-		DarkModeParams p{
-			DarkMode::isExperimentalActive() ? L"DarkMode_Explorer" : nullptr
-			, subclass
-			, theme
-		};
-
-		::EnableThemeDialogTexture(hwndParent, theme && !DarkMode::isEnabled() ? ETDT_ENABLETAB : ETDT_DISABLE);
-
-		::EnumChildWindows(hwndParent, [](HWND hWnd, LPARAM lParam) WINAPI_LAMBDA {
-			const auto& p = *reinterpret_cast<DarkModeParams*>(lParam);
-			std::wstring className = getWndClassName(hWnd);
-
-			if (className == WC_BUTTON)
-			{
-				DarkMode::subclassAndThemeButton(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_STATIC)
-			{
-				DarkMode::subclassStaticText(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_COMBOBOX)
-			{
-				DarkMode::subclassAndThemeComboBox(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_EDIT)
-			{
-				DarkMode::subclassAndThemeListBoxOrEditControl(hWnd, p, false);
-				return TRUE;
-			}
-
-			if (className == WC_LISTBOX)
-			{
-				DarkMode::subclassAndThemeListBoxOrEditControl(hWnd, p, true);
-				return TRUE;
-			}
-
-			if (className == WC_LISTVIEW)
-			{
-				DarkMode::subclassAndThemeListView(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_TREEVIEW)
-			{
-				DarkMode::themeTreeView(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == TOOLBARCLASSNAME)
-			{
-				DarkMode::themeToolbar(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == UPDOWN_CLASS)
-			{
-				DarkMode::subclassAndThemeUpDownControl(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_TABCONTROL)
-			{
-				DarkMode::subclassTabControl(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == STATUSCLASSNAME)
-			{
-				DarkMode::subclassStatusBarControl(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == WC_SCROLLBAR)
-			{
-				if (p._theme)
-				{
-					DarkMode::setDarkScrollBar(hWnd);
-				}
-				return TRUE;
-			}
-
-			if (className == WC_COMBOBOXEX)
-			{
-				DarkMode::subclassComboboxEx(hWnd, p);
-				return TRUE;
-			}
-
-			if (className == PROGRESS_CLASS)
-			{
-				if (p._theme)
-				{
-					DarkMode::subclassProgressBarControl(hWnd, p);
-				}
-				else
-				{
-					DarkMode::themeProgressBar(hWnd, p);
-				}
-				return TRUE;
-			}
-
-			// Plugin might use rich edit control version 2.0 and later
-			if (className == L"RichEdit20W" || className == L"RICHEDIT50W")
-			{
-				DarkMode::themeRichEdit(hWnd, p);
-				return TRUE;
-			}
-
-			/*
-			// for debugging
-			if (className == L"#32770")
-			{
-				return TRUE;
-			}
-
-			if (className == TRACKBAR_CLASS)
-			{
-				return TRUE;
-			}
-			*/
-
-			return TRUE;
-		}, reinterpret_cast<LPARAM>(&p));
-	}
-
-	void autoThemeChildControls(HWND hwndParent)
-	{
-		autoSubclassAndThemeChildControls(hwndParent, false, DarkMode::isWindows10());
-	}
-
-	void subclassAndThemeButton(HWND hWnd, DarkModeParams p)
+	static void subclassAndThemeButton(HWND hWnd, DarkModeParams p)
 	{
 		const auto nBtnStyle = ::GetWindowLongPtr(hWnd, GWL_STYLE);
 		switch (nBtnStyle & BS_TYPEMASK)
@@ -3808,7 +3694,7 @@ namespace DarkMode
 		}
 	}
 
-	void subclassAndThemeComboBox(HWND hWnd, DarkModeParams p)
+	static void subclassAndThemeComboBox(HWND hWnd, DarkModeParams p)
 	{
 		const auto nStyle = ::GetWindowLongPtr(hWnd, GWL_STYLE);
 
@@ -3834,15 +3720,14 @@ namespace DarkMode
 				}
 			}
 
-			if (p._theme && DarkMode::isExperimentalSupported())
+			if (p._theme)
 			{
-				DarkMode::allowDarkModeForWindow(hWnd, DarkMode::isExperimentalActive());
-				::SetWindowTheme(hWnd, L"CFD", nullptr);
+				DarkMode::setDarkThemeExperimental(hWnd, L"CFD");
 			}
 		}
 	}
 
-	void subclassAndThemeListBoxOrEditControl(HWND hWnd, DarkModeParams p, bool isListBox)
+	static void subclassAndThemeListBoxOrEditControl(HWND hWnd, DarkModeParams p, bool isListBox)
 	{
 		const auto nStyle = ::GetWindowLongPtr(hWnd, GWL_STYLE);
 		const bool hasScrollBar = ((nStyle & WS_HSCROLL) == WS_HSCROLL) || ((nStyle & WS_VSCROLL) == WS_VSCROLL);
@@ -3872,7 +3757,7 @@ namespace DarkMode
 		}
 	}
 
-	void subclassAndThemeListView(HWND hWnd, DarkModeParams p)
+	static void subclassAndThemeListView(HWND hWnd, DarkModeParams p)
 	{
 		if (p._theme)
 		{
@@ -3892,7 +3777,7 @@ namespace DarkMode
 		}
 	}
 
-	void themeTreeView(HWND hWnd, DarkModeParams p)
+	static void themeTreeView(HWND hWnd, DarkModeParams p)
 	{
 		TreeView_SetTextColor(hWnd, DarkMode::getViewTextColor());
 		TreeView_SetBkColor(hWnd, DarkMode::getViewBackgroundColor());
@@ -3906,7 +3791,7 @@ namespace DarkMode
 		}
 	}
 
-	void themeToolbar(HWND hWnd, DarkModeParams p)
+	static void themeToolbar(HWND hWnd, DarkModeParams p)
 	{
 		DarkMode::setDarkLineAbovePanelToolbar(hWnd);
 
@@ -3916,16 +3801,7 @@ namespace DarkMode
 		}
 	}
 
-	void themeRichEdit(HWND hWnd, DarkModeParams p)
-	{
-		if (p._theme)
-		{
-			//dark scrollbar for rich edit control
-			::SetWindowTheme(hWnd, p._themeClassName, nullptr);
-		}
-	}
-
-	void themeProgressBar(HWND hWnd, DarkModeParams p)
+	static void themeProgressBar(HWND hWnd, DarkModeParams p)
 	{
 		if (p._theme)
 		{
@@ -3939,6 +3815,152 @@ namespace DarkMode
 		}
 	}
 
+	static void themeRichEdit(HWND hWnd, DarkModeParams p)
+	{
+		if (p._theme)
+		{
+			//dark scrollbar for rich edit control
+			::SetWindowTheme(hWnd, p._themeClassName, nullptr);
+		}
+	}
+
+	void autoSubclassAndThemeChildControls(HWND hwndParent, bool subclass, bool theme)
+	{
+		DarkModeParams p{
+			DarkMode::isExperimentalActive() ? L"DarkMode_Explorer" : nullptr
+			, subclass
+			, theme
+		};
+
+		::EnableThemeDialogTexture(hwndParent, theme && !DarkMode::isEnabled() ? ETDT_ENABLETAB : ETDT_DISABLE);
+
+		::EnumChildWindows(hwndParent, [](HWND hWnd, LPARAM lParam) WINAPI_LAMBDA {
+			const auto& p = *reinterpret_cast<DarkModeParams*>(lParam);
+			std::wstring className = getWndClassName(hWnd);
+
+			if (className == WC_BUTTON)
+			{
+				DarkMode::subclassAndThemeButton(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_STATIC)
+			{
+				DarkMode::subclassStaticText(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_COMBOBOX)
+			{
+				DarkMode::subclassAndThemeComboBox(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_EDIT)
+			{
+				DarkMode::subclassAndThemeListBoxOrEditControl(hWnd, p, false);
+				return TRUE;
+			}
+
+			if (className == WC_LISTBOX)
+			{
+				DarkMode::subclassAndThemeListBoxOrEditControl(hWnd, p, true);
+				return TRUE;
+			}
+
+			if (className == WC_LISTVIEW)
+			{
+				DarkMode::subclassAndThemeListView(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_TREEVIEW)
+			{
+				DarkMode::themeTreeView(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == TOOLBARCLASSNAME)
+			{
+				DarkMode::themeToolbar(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == UPDOWN_CLASS)
+			{
+				DarkMode::subclassAndThemeUpDownControl(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_TABCONTROL)
+			{
+				DarkMode::subclassTabControl(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == STATUSCLASSNAME)
+			{
+				DarkMode::subclassStatusBarControl(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == WC_SCROLLBAR)
+			{
+				if (p._theme)
+				{
+					DarkMode::setDarkScrollBar(hWnd);
+				}
+				return TRUE;
+			}
+
+			if (className == WC_COMBOBOXEX)
+			{
+				DarkMode::subclassComboboxExControl(hWnd, p);
+				return TRUE;
+			}
+
+			if (className == PROGRESS_CLASS)
+			{
+				if (p._theme)
+				{
+					DarkMode::subclassProgressBarControl(hWnd, p);
+				}
+				else
+				{
+					DarkMode::themeProgressBar(hWnd, p);
+				}
+				return TRUE;
+			}
+
+			// Plugin might use rich edit control version 2.0 and later
+			if (className == L"RichEdit20W" || className == L"RICHEDIT50W")
+			{
+				DarkMode::themeRichEdit(hWnd, p);
+				return TRUE;
+			}
+
+			/*
+			// for debugging
+			if (className == L"#32770")
+			{
+				return TRUE;
+			}
+
+			if (className == TRACKBAR_CLASS)
+			{
+				return TRUE;
+			}
+			*/
+
+			return TRUE;
+		}, reinterpret_cast<LPARAM>(&p));
+	}
+
+	void autoThemeChildControls(HWND hwndParent)
+	{
+		autoSubclassAndThemeChildControls(hwndParent, false, DarkMode::isWindows10());
+	}
+
 	static LRESULT darkToolBarNotifyCustomDraw(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
 	{
 		auto lptbcd = reinterpret_cast<LPNMTBCUSTOMDRAW>(lParam);
@@ -4926,6 +4948,15 @@ namespace DarkMode
 		}
 	}
 
+	void setDarkThemeExperimental(HWND hWnd, const wchar_t* themeClassName)
+	{
+		if (DarkMode::isExperimentalSupported())
+		{
+			DarkMode::allowDarkModeForWindow(hWnd, DarkMode::isExperimentalActive());
+			::SetWindowTheme(hWnd, themeClassName, nullptr);
+		}
+	}
+
 	void disableVisualStyle(HWND hWnd, bool doDisable)
 	{
 		if (doDisable)
@@ -4940,6 +4971,29 @@ namespace DarkMode
 
 	// range to determine when it should be better to use classic style
 	constexpr double g_middleGrayRange = 2.0;
+	static TreeViewStyle g_treeViewStylePrev = TreeViewStyle::classic;
+	static COLORREF g_treeViewBg = RGB(41, 49, 52);
+	static double g_lightnessTreeView = 50.0;
+
+	// adapted from https://stackoverflow.com/a/56678483
+	double calculatePerceivedLightness(COLORREF clr)
+	{
+		auto linearValue = [](double colorChannel) -> double {
+			colorChannel /= 255.0;
+			if (colorChannel <= 0.04045)
+				return colorChannel / 12.92;
+			return std::pow(((colorChannel + 0.055) / 1.055), 2.4);
+			};
+
+		double r = linearValue(static_cast<double>(GetRValue(clr)));
+		double g = linearValue(static_cast<double>(GetGValue(clr)));
+		double b = linearValue(static_cast<double>(GetBValue(clr)));
+
+		double luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;
+
+		double lightness = (luminance <= 216.0 / 24389.0) ? (luminance * 24389.0 / 27.0) : (std::pow(luminance, (1.0 / 3.0)) * 116.0 - 16.0);
+		return lightness;
+	}
 
 	void calculateTreeViewStyle()
 	{
diff --git a/DarkMode/DarkModeSubclass.h b/DarkMode/DarkModeSubclass.h
index 076ae32ed..1ea75d02d 100644
--- a/DarkMode/DarkModeSubclass.h
+++ b/DarkMode/DarkModeSubclass.h
@@ -51,13 +51,6 @@ namespace DarkMode
 		COLORREF headerEdge = 0;
 	};
 
-	struct DarkModeParams
-	{
-		const wchar_t* _themeClassName = nullptr;
-		bool _subclass = false;
-		bool _theme = false;
-	};
-
 	enum class ToolTipsType
 	{
 		tooltip,
@@ -67,22 +60,23 @@ namespace DarkMode
 		tabbar
 	};
 
-	enum ColorTone {
-		blackTone  = 0,
-		redTone    = 1,
-		greenTone  = 2,
-		blueTone   = 3,
-		purpleTone = 4,
-		cyanTone   = 5,
-		oliveTone  = 6,
-		customizedTone = 32
+	enum class ColorTone
+	{
+		black       = 0,
+		red         = 1,
+		green       = 2,
+		blue        = 3,
+		purple      = 4,
+		cyan        = 5,
+		olive       = 6,
+		customized  = 32
 	};
 
 	enum class TreeViewStyle
 	{
 		classic = 0,
-		light = 1,
-		dark = 2
+		light   = 1,
+		dark    = 2
 	};
 
 	void initDarkMode();
@@ -97,10 +91,38 @@ namespace DarkMode
 	bool isWindows11();
 	DWORD getWindowsBuildNumber();
 
-	double calculatePerceivedLightness(COLORREF c);
+	// handle events
+	bool handleSettingChange(LPARAM lParam);
+	bool isDarkModeReg();
+
+	// from DarkMode.h
+	void setSysColor(int nIndex, COLORREF color);
+	bool hookSysColor();
+	void unhookSysColor();
+
+	// enhancements to DarkMode.h
+	void enableDarkScrollBarForWindowAndChildren(HWND hWnd);
 
 	void setDarkCustomColors(ColorTone colorTone);
 
+	COLORREF setBackgroundColor(COLORREF clrNew);
+	COLORREF setCtrlBackgroundColor(COLORREF clrNew);
+	COLORREF setHotBackgroundColor(COLORREF clrNew);
+	COLORREF setDlgBackgroundColor(COLORREF clrNew);
+	COLORREF setErrorBackgroundColor(COLORREF clrNew);
+
+	COLORREF setTextColor(COLORREF clrNew);
+	COLORREF setDarkerTextColor(COLORREF clrNew);
+	COLORREF setDisabledTextColor(COLORREF clrNew);
+	COLORREF setLinkTextColor(COLORREF clrNew);
+
+	COLORREF setEdgeColor(COLORREF clrNew);
+	COLORREF setHotEdgeColor(COLORREF clrNew);
+	COLORREF setDisabledEdgeColor(COLORREF clrNew);
+
+	void changeCustomTheme(const Colors& colors);
+	void updateBrushesAndPens();
+
 	COLORREF getBackgroundColor();
 	COLORREF getCtrlBackgroundColor();
 	COLORREF getHotBackgroundColor();
@@ -131,31 +153,31 @@ namespace DarkMode
 	HPEN getHotEdgePen();
 	HPEN getDisabledEdgePen();
 
-	void changeCustomTheme(const Colors& colors);
+	COLORREF setViewBackgroundColor(COLORREF clrNew);
+	COLORREF setViewTextColor(COLORREF clrNew);
+	COLORREF setViewGridlinesColor(COLORREF clrNew);
+
+	COLORREF setHeaderBackgroundColor(COLORREF clrNew);
+	COLORREF setHeaderHotBackgroundColor(COLORREF clrNew);
+	COLORREF setHeaderTextColor(COLORREF clrNew);
+
+	void updateBrushesAndPensView();
 
 	COLORREF getViewBackgroundColor();
 	COLORREF getViewTextColor();
 	COLORREF getViewGridlinesColor();
-	HBRUSH getViewBackgroundBrush();
-	HBRUSH getViewGridlinesBrush();
 
-	// handle events
-	bool handleSettingChange(LPARAM lParam);
-	bool isDarkModeReg();
+	COLORREF getHeaderBackgroundColor();
+	COLORREF getHeaderHotBackgroundColor();
+	COLORREF getHeaderTextColor();
 
-	// from DarkMode.h
-	void initExperimentalDarkMode();
-	void setDarkMode(bool useDark, bool fixDarkScrollbar);
-	void allowDarkModeForApp(bool allow);
-	bool allowDarkModeForWindow(HWND hWnd, bool allow);
-	void setTitleBarThemeColor(HWND hWnd);
+	HBRUSH getViewBackgroundBrush();
+	HBRUSH getViewGridlinesBrush();
 
-	void setSysColor(int nIndex, COLORREF color);
-	bool hookSysColor();
-	void unhookSysColor();
+	HBRUSH getHeaderBackgroundBrush();
+	HBRUSH getHeaderHotBackgroundBrush();
 
-	// enhancements to DarkMode.h
-	void enableDarkScrollBarForWindowAndChildren(HWND hWnd);
+	HPEN getHeaderEdgePen();
 
 	void paintRoundRect(HDC hdc, const RECT rect, const HPEN hpen, const HBRUSH hBrush, int width = 0, int height = 0);
 	inline void paintRoundFrameRect(HDC hdc, const RECT rect, const HPEN hpen, int width = 0, int height = 0);
@@ -166,22 +188,12 @@ namespace DarkMode
 	void subclassTabControlUpDown(HWND hWnd);
 	void subclassTabControl(HWND hWnd);
 	void subclassComboBoxControl(HWND hWnd);
+	void subclassComboboxExControl(HWND hWnd);
 	void subclassListViewControl(HWND hWnd);
+	void subclassHeaderControl(HWND hWnd);
 	void subclassStatusBarControl(HWND hWnd);
 	void subclassProgressBarControl(HWND hWnd);
-
-	void subclassAndThemeButton(HWND hWnd, DarkModeParams p);
-	void subclassAndThemeComboBox(HWND hWnd, DarkModeParams p);
-	void subclassAndThemeListBoxOrEditControl(HWND hWnd, DarkModeParams p, bool isListBox);
-	void subclassAndThemeListView(HWND hWnd, DarkModeParams p);
-	void themeTreeView(HWND hWnd, DarkModeParams p);
-	void themeToolbar(HWND hWnd, DarkModeParams p);
-	void themeRichEdit(HWND hWnd, DarkModeParams p);
-	void themeProgressBar(HWND hWnd, DarkModeParams p);
-	void subclassTabControl(HWND hWnd, DarkModeParams p);
-	void subclassStatusBarControl(HWND hWnd, DarkModeParams p);
-	void subclassProgressBarControl(HWND hWnd, DarkModeParams p);
-	void subclassComboboxEx(HWND hWnd, DarkModeParams p);
+	void subclassStaticText(HWND hWnd);
 
 	void autoSubclassAndThemeChildControls(HWND hWndParent, bool subclass = true, bool theme = true);
 	void autoThemeChildControls(HWND hWndParent);
@@ -197,8 +209,10 @@ namespace DarkMode
 	void setDarkTooltips(HWND hWnd, ToolTipsType type = ToolTipsType::tooltip);
 	void setDarkLineAbovePanelToolbar(HWND hWnd);
 	void setDarkListView(HWND hWnd);
+	void setDarkThemeExperimental(HWND hWnd, const wchar_t* themeClassName = L"Explorer");
 
 	void disableVisualStyle(HWND hWnd, bool doDisable);
+	double calculatePerceivedLightness(COLORREF clr);
 	void calculateTreeViewStyle();
 	void updatePrevTreeViewStyle();
 	TreeViewStyle getTreeViewStyle();
